### PR TITLE
fix(inward): name the other patient in a room overlap warning

### LINKED ISSUE #23641: Inward: room overlap warning must name the other patient and never count retired room stays
## Summary

On **Inward → Patient Room Details** (`inward/inward_patient_room_details.xhtml`) the "Room Time Overlap Detected" warning is unusable in the one case it is designed for: when the conflicting room stay belongs to a **different patient** occupying the same bed. The message names only the room, so staff see the room they are already looking at listed as its own conflict, conclude the warning is a bug, and click through it.

Separately, the overlap logic must be made explicitly and permanently blind to **retired** room stays.

## Real case (Coop production, 2026-09-08/09)

Patient **BHT/57914** (`patientencounter` 13861747) had these active room stays — no overlap between them:

| patientroom | bed | admittedAt | dischargedAt | retired |
|---|---|---|---|---|
| 13860077 | Room 90 wm | 08-Sep 15:17 | 08-Sep 19:19 | 0 |
| 13860078 | Room 511 wm | 08-Sep 19:20 | 09-Sep 10:00 | 0 |

Yet Room 90 wm rendered a red `Overlap` tag and the page header showed **"Room Time Overlap Detected"**.

The actual conflict was a *different* patient: **BHT/57843** (`patientencounter` 13861624, `patientroom` 13859990) was recorded in the **same bed, Room 90 wm** (`roomfacilitycharge` 557757) from 05-Sep 00:48 with `dischargedAt` still `NULL` at the time.

So the warning was **correct** — Room 90 is a single room (`category` 1404) and had two patients recorded in it at once. But the tooltip only said:

> Overlaps with Room 90 wm (Active)

which reads as "Room 90 wm overlaps with Room 90 wm" and tells staff nothing they can act on. There is no way from this page to learn *which patient* is holding the bed.

## Problem 1 — the warning does not identify the other patient

`BhtSummeryController.getOverlappingRooms()` deliberately matches other stays on **the same encounter OR the same bed**:

```java
// BhtSummeryController.java (~line 2040)
jpql.append("AND (pr2.patientEncounter = :pe OR pr2.roomFacilityCharge = :rfc) ");
```

but `getOverlapDescription()` (~line 2075) renders only the room name and Active/Left:

```java
sb.append(roomName).append(pr2.getDischargedAt() == null ? " (Active)" : " (Left)");
```

### Required

When the conflicting stay belongs to a **different** `PatientEncounter`, the message must identify that patient and their stay, e.g.:

> Overlaps with **Room 90 wm** — held by **BHT/57843** (S. Perera), 05-Sep 00:48 → still in room

When the conflicting stay belongs to the **same** encounter, keep the current wording (the room name is enough — the user can see the other row).

Suggested detail for the cross-patient case:
- other patient's **BHT number**
- other patient's **name**
- the other stay's **admitted → discharged** window (or "still in room" when `dischargedAt` is null)

Open question for review: is showing another patient's name acceptable on this screen, or should it be BHT number only? BHT number alone is enough to act on; the name makes it faster to confirm. Defaulting to **BHT number + name** unless there is an objection.

## Problem 2 — retired room stays must never be considered

Retired stays are historical corrections and must never produce or contribute to an overlap warning. BHT/57914 above had two retired duplicates (13860061: 15:01→19:16, and 13860076: 18:10→19:15) which *do* overlap the live 15:17→19:19 row — if they were ever counted, every corrected room record on the system would light up as a conflict.

Current state (verified on `origin/development`):
- `InwardBeanController.fetchPatientRoomAll()` filters `pr.retired = false` — the displayed list is clean
- `BhtSummeryController.getOverlappingRooms()` JPQL starts `WHERE pr2.retired = false` — the candidate side is clean
- **but** nothing guards the *subject* room: `getOverlappingRooms(patientRoom)` / `hasOverlap(patientRoom)` will happily evaluate a retired `PatientRoom` if a caller ever passes one (e.g. a stale in-memory row after `roomChangeController.removeRoom()`, or a future caller iterating an unfiltered list)
- there is no test locking any of this in

### Required

1. Add an explicit early return in `getOverlappingRooms()` when `patientRoom.isRetired()` — retired rows never have overlaps.
2. Keep the `pr2.retired = false` predicate and add a regression test so it cannot be dropped.
3. Same for `isAnyRoomOverlapping()` — skip retired entries defensively.

## Acceptance criteria

- [ ] A cross-patient overlap tooltip names the other patient's BHT number (and name) and their stay window, not just the room name
- [ ] A same-encounter overlap keeps the existing concise wording
- [ ] The confirmation dialog on **Save Changes** shows the same enriched text
- [ ] A retired `PatientRoom` never triggers an overlap warning, whether it is the subject row or a candidate row
- [ ] Regression test covering: (a) retired candidate ignored, (b) retired subject ignored, (c) cross-patient same-bed overlap detected and described with the other BHT

## Files

- `src/main/java/com/divudi/bean/inward/BhtSummeryController.java` — `getOverlappingRooms()`, `hasOverlap()`, `getOverlapDescription()`, `isAnyRoomOverlapping()`
- `src/main/webapp/inward/inward_patient_room_details.xhtml` — row `Overlap` tag (~line 162), header tag (~line 68), `dlgRoomOverlapConfirm` wording (~line 44, currently says "for this patient or bed")

## Menu path

Menu → Inpatient → (search/select BHT) → Patient Room Details

## Note — separate data problem found while investigating

While tracing this, BHT/57843's Room 90 wm stay was closed at **09-Sep 11:11:03** (`discharged = 1`, `dischargedAt` set) with **no replacement room row created**, while that encounter is still `discharged = 0`, `roomAdmitted = 1`, and `currentPatientRoom_id` still points at the now-closed stay. That patient is in the hospital with no open room stay, so room/nursing/maintenance charges stop accruing from that moment and they will be under-billed at discharge. This is a data correction for the Coop team, not part of this issue, but it is the likely manual workaround someone applied to clear this very warning — which is another reason the warning needs to say who the other patient is.

### PR DESCRIPTION
**Pull request by buddhika75**
Closes #23641

## Decisions made without approval

Run unattended (`dev-issue-unattended`), so every judgment call is listed here for the reviewer to scan first.

1. **The other patient's name is shown, not just their BHT number.** `inward/inward_room_occupancy.xhtml` already lists BHT number *and* `patientEncounter.patient.person.name` for every occupied room, to the same inward staff — so this discloses nothing that screen does not. One-line revert in `patientNameOf(...)` if that is wrong. ([reasoning on the issue](https://github.com/hmislk/hmis/issues/23641#issuecomment-5598237991))
2. **Scope grew by one item: the tooltip never rendered at all.** `p:tag` has no `title` passthrough, so `<p:tag … title="#{…getOverlapDescription(rm)}"/>` produced a DOM node with no `title` — that description has never reached a user on any build. Fixing only the string could not have met the issue's tooltip acceptance criterion, so `p:tooltip` was added (the pattern already used in `admin/lims/investigation_format_multiple.xhtml`).
3. **Multiple conflicts are separated by `;` instead of `,`** — a cross-patient entry contains commas of its own, so a comma separator became ambiguous.
4. **No caching added**, though `hasOverlap` + the tooltip value re-query per row per phase. A memoised result that outlives a row edit would leave an Overlap tag on a conflict the user just fixed — worse than the extra queries, and not something to land unattended. Filed as **#23644** instead, with the staleness trap written up.
5. **Implemented directly rather than dispatched to `java-backend-developer` / `jsf-frontend-dev`** — one controller, one facelet, one new test; delegation would have cost more than it saved.
6. **Wiki prose extended, not rewritten.** The existing "Identifying Room Overlaps" paragraph is untouched; a new section documents the tag/banner/tooltip/dialog and the two conflict kinds. Nothing existing was contradicted or deleted.

### Self-review pass (`/code-review high`, before first push)

Run on the working tree pre-commit; it raised five findings.

| Finding | Bucket | Action |
|---|---|---|
| Stay window hardcoded `dd MMM yyyy HH:mm` + default locale, while the row's own date pickers use `shortDateTimeFormat` | in-scope bug | **Fixed** — see "Formatting" below, plus 2 new tests |
| The new retired guards are unreachable from any current production path | no change | Correct, and deliberate — see "On the retired guards" |
| Tooltip adds a second overlap query per conflicting row | out of scope | **Filed as #23644** |
| Cross-encounter PHI on a single-patient screen; suggested a `ConfigOption` gate | no change | This is what the issue asks for. Gating PHI is an access-control change, which this workflow will not make unattended — flagged for the reviewer instead |
| `persistence.xml` must not be committed | process | Left unstaged; the commit carries the CI placeholders |

## Problem

`getOverlappingRooms()` matches other stays on **the same encounter OR the same bed**, so a conflict is often a *different admission* still holding the room. The message named only the room, so that case rendered as *"Room 90 overlaps with Room 90 (Active)"* — reading as a false alarm, with no way to find the other admission from this page. In the reported case that led to the other patient's room stay being closed by hand to silence the warning, leaving an in-house patient with no open room stay and their room charges no longer accruing.

## Changes

**`BhtSummeryController`**
- `isOverlapConflict(subject, candidate)` — new `static`, facade-free predicate mirroring the JPQL clause-for-clause. Retired excluded on **both** sides.
- `getOverlappingRooms()` — early return when the *subject* row is retired; query results re-checked through the predicate above.
- `describeOverlaps(subject, overlaps, pattern)` — new `static` formatter. Same-encounter conflicts keep the terse wording; a different-patient conflict is named in full: room, BHT, patient, and that stay's window.
- `isAnyRoomOverlapping()` — skips retired entries.

**Formatting** — the window now renders with `sessionController.applicationPreference.shortDateTimeFormat`, the same preference the row's `rmAdmittedAt` / `rmDischargedAt` pickers use, under `Locale.ENGLISH`, falling back to a default if the preference is unusable. Previously a deployment on `dd/MM/yyyy hh:mm a` showed the conflict as `19:19` beside a field reading `07:19 PM`.

**`inward_patient_room_details.xhtml`** — row tag: dropped the non-functional `title`, added `p:tooltip`. Save-confirmation wording now says the conflict may be *another patient recorded in the same bed*; dialog widened for the longer message.

**`BhtSummeryControllerRoomOverlapTest`** — 15 tests: retired candidate ignored, retired subject ignored, back-to-back stays, Guardian excluded, different beds, cross-patient same-bed detected, full cross-patient wording, discharged-window wording, missing-BHT fallback, same-encounter Active/Left wording, multiple conflicts, empty/null, configured pattern honoured, unusable pattern falls back.

### On the retired guards

Worth being explicit, since the review flagged it: **retired stays were already excluded** before this PR — `InwardBeanController.fetchPatientRoomAll` filters `pr.retired = false`, and the overlap JPQL opens with `pr2.retired = false`. Confirmed live: a retired stay in the test bed, whose times overlap all three live stays, raised no warning on `development` either. So the guards are a rule made total and locked in by tests, not a behaviour fix — the issue asks for the former. The one genuinely new guard is on the *subject* row, which `hasOverlap(pr)` will accept from any caller.

## Testing

**Menu path:** *Menu → Inward → Search → Admissions → (BHT No filter) → Search → Inpatient Dashboard → Room Management → Room Details*

`mvn clean package` — **276 tests, 0 failures** (15 new).

End-to-end against local Payara + local `coop` DB. The conflict was created **through the app** (room autocomplete → Save Changes → Yes, Save Anyway), not by SQL, so the save path was exercised too. The bed ended up holding four stays — three live, one retired — making every branch observable at once:

- Page banner **Room Time Overlap Detected** renders.
- Row tooltip renders and names the conflicting admission with BHT, patient and window — previously it rendered nothing at all.
- The **retired** stay in the same bed, overlapping all three live stays, is absent from the description.
- A live stay that ended one minute before the subject began is correctly **not** reported.
- **Save Changes** raises the confirmation carrying the same text; **Yes, Save Anyway** persists (verified in the DB).
- After the formatting fix, tooltip `08 Sep 26 15:39` matches the row's own field `08 Sep 26 17:52` — same preference, same rendering.

Overlap tag with hover text naming the other admission

*Patient identifiers are replaced in both screenshots.*

## Documentation

Wiki — **[Inpatient — Patient Room Details](https://github.com/hmislk/hmis/wiki/Inpatient-Patient-Room-Details)**: new *Room Time Overlap Warnings* section covering the banner, the row tag, the two kinds of conflict with the wording for each, that withdrawn room records are ignored, and that the save warning does not block the save. Both screenshots embedded with visible captions.

`developer_docs/testing/playwright-e2e-workflow.md` §113 — `p:tag` silently drops `title`; how to spot it, and to assert the rendered DOM rather than trusting the source.

## Follow-up

- **#23644** — the overlap check re-queries per row, per phase, on this page.

## Not changed

Cancellation/return behaviour, room charge calculation, and the overlap query's own predicate. The warning remains advisory and still does not block discharge.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01LXjDBhnANtPG3GDuJ9uvay

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved room-overlap detection by excluding retired rooms and validating conflicts more reliably.
  - Clarified warnings for overlaps involving the same patient versus another patient.
  - Conflict details now display patient, BHT, and stay-time information using the configured date format.
  - Room-overlap dialogs provide clearer, more spacious warnings.
  - Room-overlap tooltips now display consistently, including within data tables.

- **Documentation**
  - Added guidance for implementing and testing room-overlap tooltips.

- **Tests**
  - Expanded coverage for overlap detection, conflict descriptions, date formats, and edge cases.

### CODE REVIEW

File: developer_docs/testing/playwright-e2e-workflow.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ waste a session.
 
 ## Contents
 
-111 sections. **The workflow is §0-§8; everything from §9 on is an independent
+118 sections. **The workflow is §0-§8; everything from §9 on is an independent
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply developer_docs/testing/playwright-e2e-workflow.md:17]: Valid — fixed in 74f565df46.

The gap was slightly wider than reported: **§111 and §112 were missing from the Contents too** (pre-existing drift from earlier PRs, not this branch). Since the fix edits that exact list, all three were added rather than just §113.

One correction to the proposed diff: it wraps the anchor in backticks — ``](`#113-...`)`` — which would render as literal text instead of a link. Used the plain form the rest of the file uses.

Verified every Contents link resolves to a real heading — **119 links, 0 broken** — under GitHub's slug rule, where the spaces either side of an em dash each become a hyphen (`title--a-tooltip`), not a single one.

Not fixed here, flagged for a future pass: the Contents has a **duplicate `107.`** (lines 137 and an earlier entry), and one entry is listed out of numeric order (`96.` sits between `108.` and `107.`). Both predate this branch and renumbering would churn anchors other docs may link to.
